### PR TITLE
release: v0.5.2

### LINKED ISSUE #26: Remove best-effort flush loop in shutdown() for close()-time buffer discard
## Context

Related: wspulse/.github#30 (cross-SDK alignment of close()-time send buffer behaviour)

The behaviour contract (wspulse/.github#29) now states that `close()` discards unsent buffered frames. The current TypeScript implementation does not comply: `shutdown()` contains a best-effort flush loop that iterates through `sendBuffer` and fires `sendOneFrame()` for each frame (without awaiting results) before calling `ws.close()`.

## Required Change

Remove the best-effort flush loop in the `shutdown()` method. On close, the send buffer should be discarded, not flushed.

### Target code (`src/client.ts`, `shutdown()` method)

Find and remove the flush loop (approximate location):

```typescript
// Best-effort flush: fire each pending frame with writeWait deadline.
// Results are intentionally not awaited — shutdown proceeds immediately.
while (this.sendBuffer.length > 0) {
  void this.sendOneFrame(this.sendBuffer.shift() as string | Uint8Array);
}
```

Replace with explicit buffer discard:

```typescript
// Discard unsent frames — close() does not drain the send buffer.
this.sendBuffer.length = 0;
```

### Effect

- **Before**: shutdown fires all buffered frames (fire-and-forget), then closes WebSocket. Frames may or may not actually be delivered depending on timing.
- **After**: shutdown discards all buffered frames immediately, then closes WebSocket. No ambiguity.

### Notes

- The `flushSendBuffer()` method (normal write loop during connected state) is unchanged — it still drains frames one by one with full deadline enforcement during normal operation.
- Review existing tests to confirm none depend on "flush on close" behaviour.
- Run `make check` to verify.

## Checklist

- [x] Remove flush loop in `shutdown()`
- [x] Add explicit buffer discard
- [x] Verify `make check` passes
- [x] CHANGELOG entry

### LINKED ISSUE #27: perf: replace O(n) array-shift send buffer with ring buffer
Parent issue: wspulse/.github#28

## Problem

`client-ts` uses a plain array as its outbound send buffer. Dequeuing the front element with `Array.prototype.shift()` is O(n) because the runtime must copy all remaining elements forward. In a serial drain loop this produces O(n^2) total work for a full buffer flush.

## Proposed solution

Replace the array-based send buffer with a `RingBuffer<string | Uint8Array>` class using modular index arithmetic:

1. **New file `src/ring_buffer.ts`** -- `RingBuffer` class with:
   - Fixed capacity (set to `sendBufferSize`)
   - O(1) `push()` and O(1) `dequeue()` via head/tail pointers
   - Semantics matching `server/resume.go`

2. **Update `src/client.ts`** -- replace `(string | Uint8Array)[]` + `push()`/`shift()` with `RingBuffer` usage in the write loop

3. **Tests** -- unit tests for `RingBuffer` (capacity, wrap-around, full/empty edge cases) and updated component tests for send buffer behaviour

## Scope

- Internal refactor only -- no API changes, no behaviour changes
- `sendBufferSize` becomes the fixed ring buffer capacity

## Reference

- `wspulse/server/resume.go` -- existing ring buffer implementation for reference
- Parent issue wspulse/.github#28 for full context and cross-SDK comparison

### PR DESCRIPTION
**Pull request by maxence2997**
## Summary

Promotes two completed features from develop to main: discard send buffer on close (contract compliance fix) and ring buffer send buffer (O(1) enqueue/dequeue).

## Related issues

Closes #27
Relates to wspulse/.github#28

## Changes

- `fix/close-buffer-discard` (PR #28): `close()` now discards unsent buffered frames instead of flushing them — aligns with the behaviour contract
- `refactor/ring-buffer-send` (PR #29): internal send buffer replaced with `RingBuffer<T>` for O(1) push and dequeue; 15 new unit tests; `T extends NonNullable<unknown>` constraint; `src/ring-buffer.ts` naming consistent with test file

### Required

- [x] `make check` passes (format → lint → type-check → test)
- [x] Each commit represents exactly one logical change
- [x] Commit messages follow the format in `commit-message-instructions.md`
- [x] No unrelated code reformatting in this PR

### CODE REVIEW

File: src/ring-buffer.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * @internal
+ */
+export class RingBuffer<T extends NonNullable<unknown>> {
+  private readonly data: (T | undefined)[];
+  private head = 0;
+  private size = 0;
+  private readonly cap: number;
+
+  constructor(capacity: number) {
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply src/ring-buffer.ts:15]: Not applicable — same finding as wspulse/client-ts#29 (review comment on `src/ring_buffer.ts`). Already responded there: `RingBuffer` is `@internal` and is only instantiated via `opts.sendBufferSize`, which is validated upstream in `validateOptions()` (`Number.isFinite` + `Number.isInteger` + range `[1, 4096]`). Adding duplicate validation in an internal class adds code for a scenario that cannot occur through the public API.

File: Makefile [hunk 2]
```diff
@@ -22,6 +22,8 @@ fmt: ## Format source files with Prettier
 	@npx prettier --write "src/**/*.ts" "test/**/*.ts"
 
 check: ## Run fmt-check, lint, and unit tests
+	@echo "── install ──"
+	@npm install --silent
 	@echo "── fmt ──"
```
[the remark this answers is not in the document]
Comment (maxence2997) [reply Makefile:27]: Won't fix. The `npm install` in `make check` was added deliberately (beea667) to improve local developer ergonomics — running `make check` on a fresh clone no longer fails with a missing-module error. `npm install` is a near no-op when `node_modules` is already present. CI is unaffected: it runs `npm ci` independently.
